### PR TITLE
Migrates nll_loss_forward from TH to Aten (CUDA)

### LINKED ISSUE #24610: Migrate `nll_loss_forward` from the TH to Aten (CUDA)
Porting TH operators is essential for code simplicity and performance reasons.

Porting guides and Q&A are available in umbrella issue: #24507

Feel free to add @VitalyFedyunin as a reviewer to get a prioritized review.

### PR DESCRIPTION
**Pull request by thomasjpfan**
Fixes https://github.com/pytorch/pytorch/issues/24610
Aten Umbrella issue https://github.com/pytorch/pytorch/issues/24507
Related to https://github.com/pytorch/pytorch/issues/59765

The performance does not change between this PR and master with the following benchmark script:

<details>
 <summary>Benchmark script</summary>

```python
import torch
import torch.nn as nn
import time

torch.manual_seed(0)


def _time():
    torch.cuda.synchronize()
    MS_PER_SECOND = 1000
    return time.perf_counter() * MS_PER_SECOND


device = "cuda"
C = 30
softmax = nn.LogSoftmax(dim=1)
n_runs = 250

for reduction in ["none", "mean", "sum"]:
    for N in [100_000, 500_000, 1_000_000]:
        fwd_t = 0
        bwd_t = 0
        data = torch.randn(N, C, device=device)
        target = torch.empty(N, dtype=torch.long, device=device).random_(0, C)
        loss = nn.NLLLoss(reduction=reduction)
        input = softmax(data)

        for i in range(n_runs):
            t1 = _time()
            result = loss(input, target)
            t2 = _time()
            fwd_t = fwd_t + (t2 - t1)
        fwd_avg = fwd_t / n_runs
        print(
            f"input size({N}, {C}), reduction: {reduction} "
            f"forward time is {fwd_avg:.2f} (ms)"
        )
    print()
```

</details>

## master

```
input size(100000, 30), reduction: none forward time is 0.02 (ms)
input size(500000, 30), reduction: none forward time is 0.08 (ms)
input size(1000000, 30), reduction: none forward time is 0.15 (ms)

input size(100000, 30), reduction: mean forward time is 1.81 (ms)
input size(500000, 30), reduction: mean forward time is 8.24 (ms)
input size(1000000, 30), reduction: mean forward time is 16.46 (ms)

input size(100000, 30), reduction: sum forward time is 1.66 (ms)
input size(500000, 30), reduction: sum forward time is 8.24 (ms)
input size(1000000, 30), reduction: sum forward time is 16.46 (ms)
```

## this PR

```
input size(100000, 30), reduction: none forward time is 0.02 (ms)
input size(500000, 30), reduction: none forward time is 0.08 (ms)
input size(1000000, 30), reduction: none forward time is 0.15 (ms)

input size(100000, 30), reduction: mean forward time is 1.80 (ms)
input size(500000, 30), reduction: mean forward time is 8.24 (ms)
input size(1000000, 30), reduction: mean forward time is 16.46 (ms)

input size(100000, 30), reduction: sum forward time is 1.66 (ms)
input size(500000, 30), reduction: sum forward time is 8.24 (ms)
input size(1000000, 30), reduction: sum forward time is 16.46 (ms)
```

### CODE REVIEW

File: aten/src/ATen/native/cuda/Loss.cu [hunk 1]
```diff
@@ -153,4 +154,290 @@ Tensor& binary_cross_entropy_backward_out_cuda(const Tensor& grad, const Tensor&
...
+    checkAllSameGPU(
+        "nll_loss_forward_out_cuda",
+        {output_arg, total_weight_arg, input_arg, target_arg});
+  }
+
+  const int64_t n_classes = input.size(-1);
+  const int64_t n_dims = input.dim();
+  TORCH_CHECK(n_dims > 0 && n_dims <= 2, "input tensor should be 1D or 2D");
+  TORCH_CHECK(
+      target.dim() == 1,
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/Loss.cu:301]: What if target.dim() is 0? Previous code errored on this too

Comment (thomasjpfan) [reply aten/src/ATen/native/cuda/Loss.cu]: I think the original implementation did not raise this specific error when `target.dim() == 0`:

https://github.com/pytorch/pytorch/blob/d5988c5eca0221e9ef58918e4f0b504940cb926a/aten/src/THCUNN/generic/ClassNLLCriterion.cu#L14-L16

I updated the check to `TORCH_CHECK(target.dim() <= 1)` to not change the original behavior.

Comment (ngimel) [reply aten/src/ATen/native/cuda/Loss.cu]: Can you please beef up testing for supported/unsupported dims?

File: aten/src/ATen/native/cuda/Loss.cu [hunk 2]
```diff
@@ -153,4 +154,290 @@ Tensor& binary_cross_entropy_backward_out_cuda(const Tensor& grad, const Tensor&
...
+  int64_t n_dims = input.dim();
+
+  TensorArg output_arg{output, "output", 1};
+  TensorArg total_weight_arg{total_weight, "total_weight", 2};
+  TensorArg input_arg{input, "input", 3};
+  TensorArg target_arg{target, "target", 4};
+
+  if (weight.defined()) {
+    TensorArg weight_arg{weight, "weight", 5};
+    checkAllSameGPU(
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Loss.cu:294]: the device placement of all the args is automatically checked in codegen, you don't have to do it here (that also means you don't have to create TensorArgs

File: aten/src/ATen/native/cuda/Loss.cu [hunk 3]
```diff
@@ -153,4 +154,290 @@ Tensor& binary_cross_entropy_backward_out_cuda(const Tensor& grad, const Tensor&
...
+  TORCH_CHECK(
+      batch_size == num_targets,
+      "size mismatch (got input: ",
+      input.sizes(),
+      ", target: ",
+      target.sizes(),
+      ")")
+
+  TORCH_CHECK(
+      !weight.defined() || weight.numel() == n_classes,
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Loss.cu:315]: should you check weight.dim() here (and also add a test case for incorrectly sized/dimensioned weights?)

File: aten/src/ATen/native/cuda/Loss.cu [hunk 4]
```diff
@@ -153,4 +154,290 @@ Tensor& binary_cross_entropy_backward_out_cuda(const Tensor& grad, const Tensor&
...
+
+    AT_DISPATCH_FLOATING_TYPES_AND2(
+        at::ScalarType::Half,
+        at::ScalarType::BFloat16,
+        input.scalar_type(),
+        "nll_loss_forward_no_reduce_cuda_kernel",
+        [&] {
+          scalar_t* weight_data = nullptr;
+          if (weight.defined()) {
+            auto weight_ = weight.contiguous();
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Loss.cu:338]: `contiguous` should be called outside of dispatch guard, and here you can just set `weight_data` with a ternary operator

File: aten/src/ATen/native/cuda/Loss.cu [hunk 5]
```diff
@@ -153,4 +154,290 @@ Tensor& binary_cross_entropy_backward_out_cuda(const Tensor& grad, const Tensor&
...
+  if (n_dims == 1) {
+    AT_DISPATCH_FLOATING_TYPES_AND2(
+        at::ScalarType::Half,
+        at::ScalarType::BFloat16,
+        input.scalar_type(),
+        "nll_loss_forward_reduce_cuda_kernel_1d",
+        [&] {
+          scalar_t* weight_data = nullptr;
+          if (weight.defined()) {
+            auto weight_ = weight.contiguous();
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Loss.cu:373]: same here, move `weight.contiguous()` to outside of dispatch macro

**Review by ngimel — COMMENTED** [review]
This looks good, I left a few comments

File: aten/src/ATen/native/cuda/Loss.cu [hunk 6]
```diff
@@ -153,4 +154,290 @@ Tensor& binary_cross_entropy_backward_out_cuda(const Tensor& grad, const Tensor&
...
+
+  if (threadIdx.x == 0) {
+    *output = *total_weight = static_cast<scalar_t>(0);
+    accscalar_t output_acc = 0;
+    accscalar_t total_weight_acc = 0;
+    for (i = 0; i < NLL_LOSS_THREADS; ++i) {
+      output_acc += sh_inputs[i];
+      total_weight_acc += acc_weight[i];
+    }
+    *total_weight = static_cast<scalar_t>(total_weight_acc);
```
[the remark this answers is not in the document]
Comment (thomasjpfan) [reply aten/src/ATen/native/cuda/Loss.cu:250]: In this case, I think `*total_weight` still needs to be set here because it is one of the outputs of the kernel.

File: aten/src/ATen/native/cuda/Loss.cu (hunk 3 above)
Comment (thomasjpfan) [reply aten/src/ATen/native/cuda/Loss.cu]: I added more tests for input validation and also made a slight adjustment to the CPU version to also check for `weight.dim`.

File: aten/src/ATen/native/cuda/Loss.cu (hunk 6 above)
[the remark this answers is not in the document]
Comment (ngimel) [reply aten/src/ATen/native/cuda/Loss.cu:250]: Ah, sorry, didn't notice that it's the output, you are right.

File: aten/src/ATen/native/cuda/Loss.cu [hunk 7]
```diff
@@ -153,4 +154,260 @@ Tensor& binary_cross_entropy_backward_out_cuda(const Tensor& grad, const Tensor&
...
+    if (size_average) {
+      if (nframe == 0) {
+        // Mean reduction on empty tensors produces NaN
+        *output = std::numeric_limits<double>::quiet_NaN();
+      }
+      if (total_weight_acc != 0) {
+        *output = static_cast<scalar_t>(output_acc / total_weight_acc);
+      }
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Loss.cu:258]: you need `else` here, otherwise `output` won't be set for total_weight_acc = 0, and please add tests for all these cases.

Comment (thomasjpfan) [reply aten/src/ATen/native/cuda/Loss.cu]: The original implementation had `*output` initialized to 0, which happened to be correct when `total_weight_acc==0`.

I redid the implementation a little to only set `*output` once and added test for the `nan` case and the `size_average && total_weight_acc == 0` case.

**Comment by ngimel** [comment]
Test errors are real

**Review by ngimel — APPROVED** [review]

**Comment by thomasjpfan** [comment]
While looking at a [test error](https://app.circleci.com/pipelines/github/pytorch/pytorch/340276/workflows/8b476edc-b515-44b2-8420-2e4b8171220a/jobs/14312737) I saw that we already have test for zero weights and empty tensors: `test_nll_loss_empty_tensor_reduction_*` and `test_nll_loss_total_weight_is_zero`, so I removed the tests I added in this PR that tested the same thing.

The rest of the public tests seems unrelated to this PR.
